larpandoracontent/LArControlFlow/StitchingCosmicRayMergingTool.cc

```diff
@@ -764,41 +674,81 @@ void StitchingCosmicRayMergingTool::CalculateX0(const PfoToLArTPCMap &pfoToLArTP
+                    }
                 }
 
-                PfoToPointingVertexMap::iterator pfoToPointingVertexMapIter2(pfoToPointingVertexMap.find(pPfo2));
-                if (pfoToPointingVertexMapIter2 == pfoToPointingVertexMap.end())
+                // Record pfo2 stitching vertex for pfo1<->pfo2 match, used to determine shifting direction in later step
+                const PfoToPointingVertexMatrix::iterator pfoToPointingVertexMatrixIter2(pfoToPointingVertexMatrix.find(pPfo2));
+                if (pfoToPointingVertexMatrixIter2 == pfoToPointingVertexMatrix.end())
                 {
-                    (void) pfoToPointingVertexMap.insert(PfoToPointingVertexMap::value_type(pPfo2, pointingVertex2));
+                    // No matches present in map, add pfo1<->pfo2 match
+                    const PfoToPointingVertexMap pfoToPointingVertexMap({{pPfo1, pointingVertex2}});
+                    (void) pfoToPointingVertexMatrix.insert(PfoToPointingVertexMatrix::value_type(pPfo2, pfoToPointingVertexMap));
                 }
                 else
                 {
-                    if ((pfoToPointingVertexMapIter2->second.GetPosition() - pointingVertex2.GetPosition()).GetMagnitude() > std::numeric_limits<float>::epsilon())
-                        throw StatusCodeException(STATUS_CODE_FAILURE);;
+                    // ATTN: another match for a different TPC boundary may be present, add pfo1<->pfo2 match                    
+                    PfoToPointingVertexMap &pfoToPointingVertexMap(pfoToPointingVertexMatrixIter2->second);
+                    const PfoToPointingVertexMap::iterator pfoToPointingVertexMapIter(pfoToPointingVertexMap.find(pPfo1));
+                    if(pfoToPointingVertexMapIter == pfoToPointingVertexMap.end())
+                    {
+                        (void) pfoToPointingVertexMap.insert(PfoToPointingVertexMap::value_type(pPfo1, pointingVertex2));
+                    }
+                    else
+                    {
+                        if ((pfoToPointingVertexMapIter->second.GetPosition() - pointingVertex2.GetPosition()).GetMagnitude() > std::numeric_limits<float>::epsilon())
+                            throw StatusCodeException(STATUS_CODE_FAILURE);;
+                    }
+                }
+
+                const float tpcBoundaryCenterX(LArStitchingHelper::GetTPCBoundaryCenterX(*pLArTPC1, *pLArTPC2));
+                const bool isCPAStitch(pLArTPC1->GetCenterX() < tpcBoundaryCenterX ? !pLArTPC1->IsDriftInPositiveX() : !pLArTPC2->IsDriftInPositiveX());
+                float thisX0(LArStitchingHelper::CalculateX0(*pLArTPC1, *pLArTPC2, pointingVertex1, pointingVertex2));
+
+                thisX0 *= isCPAStitch ? -1.f : 1.f;
+
+                // Check if stitching contribution is consistent
+                if (sumN > std::numeric_limits<float>::epsilon())
+                {
+                    float fractionalDiff(std::fabs((sumX - (thisX0 * sumN)) / sumX));
```

Comment:
Hi Isobel, Thanks again for all this great work! This is the only issue I've identified in the latest commits, which I otherwise have in a nicely rebased and whitespace-removed branch, waiting for merge:

What should we do if sumX is effectively zero here? I not that the logic is somewhat subtle, in that sumX and sumN are incremented below.

Should we actually be asking to compare a previous running meanX0 (`sumX / sumN`) vs. the x0 associated with this new boundary (`thisX0`), and look to bail if they have a certain fractional disagreement (or even absolute disagreement if meanX0 is ~zero)? Let's check this, and then the branch can go in.

(My initial thinking here was that, as ever, if there's a divide by a value, should check that it's not zero. Then that got me thinking about what this quantity actually is.)

Thanks again!